Allows customizing the ObjectMapper in REST Client Reactive Jackson

Sgitario opened the pull request:
The REST Client Reactive supports adding a custom ObjectMapper to be used only the Client using the annotation `@ClientObjectMapper`. 

A simple example is to provide a custom ObjectMapper to the REST Client Reactive Jackson extension by doing:

```java
@Path("/extensions")
@RegisterRestClient
public interface ExtensionsService {

    @GET
    Set<Extension> getById(@QueryParam("id") String id);

    @ClientObjectMapper <1>
    static ObjectMapper objectMapper(ObjectMapper defaultObjectMapper) { <2>
        return defaultObjectMapper.copy()
                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
                .disable(DeserializationFeature.UNWRAP_ROOT_VALUE);
    }
}
```

<1> The method must be annotated with `@ClientObjectMapper`.
<2> It's must be a static method. Also, the parameter `defaultObjectMapper` will be resolved via CDI. If not found, it will throw an exception at runtime.

Closes https://github.com/quarkusio/quarkus/issues/23979

Sgitario commented:
@geoand let me know what you think about this solution:
- it follows the same pattern as what was done for `@ClientExceptionMapper` or `@ClientRedirectHandler`
- instead of providing a specific solution for Object Mappers, I thought it was a better idea to find a general solution for any kind of service by adding `@ClientProvider` which is fully integrated by REST Client Jackson and the Providers interface (I've also updated the documentation to guide users how to provide a custom object mapper per client)

geoand commented:
To be honest, I don't see why would call the annotation `@ClientProvider` and give the illusion that it's generic.
For Jakarta REST `Provider ` classes, users can already use `@RegisterProvider`, so I think the name is very misleading.

Sgitario commented:
> Provider

`@ClientProvider` is equal to `@Provider` but for the client only and for methods. Any alternative name that came up to you?

geoand commented:
What I am saying is that I prefer we have a specific annotation, just like we do (as you pointed out) for `@ClientExceptionMapper` and others

Sgitario commented:
> What I am saying is that I prefer we have a specific annotation, just like we do (as you pointed out) for `@ClientExceptionMapper` and others.
> 
> Also keep in mind, that the name `Provider` means pretty much nothing to most people (let's not forget that most people never have to use them in Jakarta REST)

Do you mean to reduce the scope of the feature to only support object mappers? And so, provide a specific annotation `@ClientObjectMapper`. 

If so, I encounter the current solution as much more powerful. The existing annotation allows adding beans to the client context in a generic way. Though, we could discuss the name (perhaps `@RegisterContextResolver` or sth similar). 

I also like having something `@ClientObjectMapper`, but without reducing the scope of these changes. This is having this annotation as a synonym of the `@ClientContextResolver` annotation (or how we agree to name it).

geoand commented:
> Do you mean to reduce the scope of the feature to only support object mappers? And so, provide a specific annotation @ClientObjectMapper

Yes

> If so, I encounter the current solution as much more powerful. The existing annotation allows adding beans to the client context in a generic way. Though, we could discuss the name (perhaps @RegisterContextResolver or sth similar

The problem is that these non-Quarkus APIs are very arcance to use (hence the original issue in the first place). 
As these are not APIs we would have come up with if we were starting from scratch, I don't really want to have more way to expose these. 

Having said that, if you want to keep the "infrastructure" you created to support it and simply not "expose" it, that's fine with me (as long as the design decision is mentioned properly in the code comments)

Sgitario commented:
> > Do you mean to reduce the scope of the feature to only support object mappers? And so, provide a specific annotation @ClientObjectMapper
> 
> Yes
> 
> > If so, I encounter the current solution as much more powerful. The existing annotation allows adding beans to the client context in a generic way. Though, we could discuss the name (perhaps @RegisterContextResolver or sth similar
> 
> The problem is that these non-Quarkus APIs are very arcance to use (hence the original issue in the first place). As these are not APIs we would have come up with if we were starting from scratch, I don't really want to have more way to expose these.
> 

What non-Quarkus APIs do you mean? The Providers interface and ContextResolver are coming from the JAX-RS specification, so the more we are integrated with it, the better (because users already know it). 

Moreover, we're documenting this use case, so from a user's perspective, it does __exist__ and it's up to the users, to use it or not.

geoand commented:
> What non-Quarkus APIs do you mean? The Providers interface and ContextResolver are coming from the JAX-RS specification

These are what I am talking about

> so the more we are integrated with it, the better (because users already know it)

We already have integration with them. What I am saying that is they are terrible APIs and we should not be trying to make things easier for users by exposing them more.

Sgitario commented:
> > What non-Quarkus APIs do you mean? The Providers interface and ContextResolver are coming from the JAX-RS specification
> 
> These are what I am talking about
> 
> > so the more we are integrated with it, the better (because users already know it)
> 
> We already have integration with them. What I am saying that is they are terrible APIs and we should not be trying to make things easier for users by exposing them more.

This is a biased argument ;)

But yes, I agree with you that having to do something like this `providers.getContextResolver(ObjectMapper.class, MediaType.APPLICATION_JSON).getContext(null);` is not great. 

I've already updated the solution and will update the pull request once all the checks are green in my local.

geoand commented:
> This is a biased argument ;)

It is :). But I also get to see 10s of issues, chats, SO questions and reproducers everyday so it's not a totally uninformed opinion :)

Sgitario commented:
PR updated and PR title/description updated @geoand

geoand commented on docs/src/main/asciidoc/rest-client-reactive.adoc:
```diff
@@ -674,6 +674,10 @@ public class ExtensionsResource {
 
 <1> the client will use the registered redirect handler over the redirect handler provided via CDI if any.
 
+== Client Context Providers
+
+
+
```
This probably should be removed :)

Sgitario replied:
ups

geoand commented on docs/src/main/asciidoc/rest-client-reactive.adoc:
```diff
@@ -1101,6 +1105,33 @@ public class TestClientRequestFilter implements ResteasyReactiveClientRequestFil
+
+    @ClientObjectMapper <1>
+    static ObjectMapper objectMapper(ObjectMapper defaultObjectMapper) { <2>
+        return defaultObjectMapper.copy()
```
We should mention here that users should `NEVER` change the `defaultObjectMapper` but instead always use `copy` if they pan to inherit the default settings. 
The same goes for the Javadoc of `@ClientObjectMapper`

Sgitario replied:
Added

geoand commented on extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/ClientContextResolverHandler.java:
```diff
@@ -0,0 +1,180 @@
+ * Generates an implementation of {@link ResteasyReactiveContextResolver}
+ * from an instance of {@link jakarta.ws.rs.ext.Provider}
+ */
+class ClientContextResolverHandler {
```
I would like some more general comments about this and why it's done like this. 
Asking because in a month, I am pretty sure no one will remember why we did it this way

Sgitario replied:
added

geoand replied:
Are you sure you have pushed the change?

Sgitario replied:
Now yes

geoand commented on extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/ResteasyReactiveContextResolver.java:
```diff
@@ -0,0 +1,6 @@
+
+import jakarta.ws.rs.ext.ContextResolver;
+
+public interface ResteasyReactiveContextResolver<T> extends ContextResolver<T> {
```
Let's add some Javadoc to this too just explaining it's functionality

Sgitario replied:
Added

Sgitario commented:
PR updated with your comments.

geoand commented on extensions/resteasy-reactive/rest-client-reactive-jackson/runtime/src/main/java/io/quarkus/rest/client/reactive/jackson/ClientObjectMapper.java:
```diff
@@ -0,0 +1,46 @@
+ * {@code
+ * &#64;ClientObjectMapper
+ * static ObjectMapper objectMapper(ObjectMapper defaultObjectMapper) {
+ *     return defaultObjectMapper.copy();
```
It probably makes sense to also change something here.

Sgitario replied:
Yes, it makes sense. Added.

Sgitario committed:
Allows customizing the ObjectMapper in REST Client Reactive Jackson

The REST Client Reactive supports adding a custom ObjectMapper to be used only the Client using the annotation `@ClientObjectMapper`. 

A simple example is to provide a custom ObjectMapper to the REST Client Reactive Jackson extension by doing:

```java
@Path("/extensions")
@RegisterRestClient
public interface ExtensionsService {

    @GET
    Set<Extension> getById(@QueryParam("id") String id);

    @ClientObjectMapper <1>
    static ObjectMapper objectMapper(ObjectMapper defaultObjectMapper) { <2>
        return defaultObjectMapper.copy()
                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
                .disable(DeserializationFeature.UNWRAP_ROOT_VALUE);
    }
}
```

<1> The method must be annotated with `@ClientObjectMapper`.
<2> It's must be a static method. Also, the parameter `defaultObjectMapper` will be resolved via CDI. If not found, it will throw an exception at runtime.

Fix https://github.com/quarkusio/quarkus/issues/23979

Sgitario commented:
PR updated again with your latest comments.

geoand reviewed (approved):
Thanks!
